File: apps/animals-api/core/contributions.py
"""What the public contribute flow shares across the three things it can create.

``apps/animals`` grew a public, staged "add a record" flow for signed-in readers
- a FAB on a category page proposes a **species**, one on a sighting page files a
**sighting**, and an inline panel in that form adds a **place** - and every one of
them needs the same three things:

* **a photo list**, uploaded as base64 in one request with the record itself,
  rather than the CMS's write-the-gallery-after-save dance - a contributor has
  one Submit button, and a half-created record with no pictures is worse than a
  failed one;
* **a hard ceiling** on how much one anonymous-ish request may cost; and
* **the same moderation contract**: the row is created ``enabled=False``, marked
  ``is_contribution``, stamped with ``created_by``, and may not set any field an
  administrator owns.

The module has **two halves**. Everything above ``ContributionUpdateSerializer``
is the create path; everything below it is the edit path that ``apps/animals``'
"My contributions" page reads and writes, which is a later and wider grant - a
contributor may now list, correct and delete what they filed. See
``core/my_contributions.py`` for the views, and the section comment on the edit
half for the two rules editing adds.

Everything here is deliberately narrow. The contribute serializers do **not**
subclass the CMS write serializers, and the edit serializers do not subclass the
create ones, because inheriting a field list is exactly how ``is_featured`` or
``enabled`` would one day become writable by the public: the field lists are
separate on purpose, and the only way to widen a public one is to type the field
into it.
"""

from django.db import transaction
from rest_framework import serializers

from .image_sizes import photo_cfg
from .serializers import ImageProcessingSerializer

__all__ = [
    'MAX_CONTRIBUTION_PHOTOS',
    'MAX_TEXT_LENGTH',
    'ContributionSerializer',
    'ContributionUpdateSerializer',
    'photos_field',
    'photos_patch_field',
]

# One outing's worth of pictures. The cap is about cost, not taste: every photo
# in this payload is decoded and resized in-process, so an uncapped list is a way
# to hold a worker for as long as the request body is allowed to be.
MAX_CONTRIBUTION_PHOTOS = 10

# A generous field-note, still short of a request that exists to fill the column.
MAX_TEXT_LENGTH = 5000


def photos_field(required=True):
    """The ``photos`` field: base64 data URLs, in the order they should appear.

    Order is the whole payload's meaning beyond the pictures themselves - the API
    publishes a record's **first** gallery row as its cover (see this app's
    CLAUDE.md), so ``photos[0]`` is the contributor's chosen cover.
    """
    return serializers.ListField(
        child=serializers.CharField(),
        required=required,
        allow_empty=not required,
        min_length=1 if required else 0,
        max_length=MAX_CONTRIBUTION_PHOTOS,
    )


class ContributionSerializer(serializers.ModelSerializer):
    """Base for the two public create serializers.

    A subclass declares its own ``Meta.fields`` (never inherited - see the module
    docstring), sets ``photo_write_serializer_class`` to the gallery writer for
    its parent, and implements nothing else: ``create`` stamps the moderation
    fields and writes the photos.
    """

    #: The existing gallery write serializer for this record's photo table. Its
    #: ``save(parent)`` is what turns one base64 string into a stored row, so the
    #: contribute path shares the CMS's image pipeline rather than a second copy.
    photo_write_serializer_class = None

    def validate_photos(self, value):
        """Reject an undecodable photo *before* the parent row is created.

        Without this the first bad data URL in a list of eight would fail halfway
        through writing the gallery, leaving a published-nowhere record with three
        of its photos. Validation up front, writes after the row exists.
        """
        for index, raw in enumerate(value):
            # The same tier the gallery writer below stores at - this pass only
            # decides whether the payload decodes, so a smaller box here would
            # reject nothing but would make the two halves read as different
            # rules.
            sub = ImageProcessingSerializer(
                data={'base64_image': raw}, **photo_cfg()
            )
            if not sub.is_valid():
                raise serializers.ValidationError(
                    {str(index): sub.errors['base64_image']}
                )
        return value

    def _request_user(self):
        request = self.context.get('request')
        user = getattr(request, 'user', None)
        return user if (user is not None and user.is_authenticated) else None

    def _write_photos(self, instance, photos):
        writer = self.photo_write_serializer_class
        for index, raw in enumerate(photos):
            sub = writer(data={'image': raw, 'sort_order': index})
            # Already validated in `validate_photos`; this second pass is the
            # writer's own field coercion, not a re-check of the payload.
            sub.is_valid(raise_exception=True)
            sub.save(instance)


# ---------------------------------------------------------------------------
# Editing a contribution back
# ---------------------------------------------------------------------------
#
# The create serializers above were, for a long time, the whole of this feature:
# a contributor filed a record and never saw it again. `apps/animals`' "My
# contributions" page is what changed that, and it needs the same three records
# to be **readable, editable and deletable by the account that filed them** -
# which is a wider grant than creating, and so is written with the same
# separate-serializer discipline rather than by relaxing anything.
#
# Two rules the update path adds, and neither belongs to a single model:
#
# * **Editing a published contribution takes it off the site.** `enabled` is not
#   writable here (it never is on a contribute serializer), but a PATCH from its
#   author flips it to False, so a live record returns to the review queue rather
#   than the public seeing an unreviewed edit. `was_published` is what then tells
#   that state apart from a first submission - see `core.models`.
# * **The photo list is a diff, not an append.** See `photos_patch_field`.


class PhotoPatchItem(serializers.Serializer):
    """One entry in a `photos` diff: either a row to keep, or a photo to add.

    Exactly one of the two, and never both - an item carrying an `id` *and* an
    `image` would be asking to replace a stored photograph in place, which is a
    thing this flow deliberately cannot do: the gallery row's filename embeds its
    pk and its `sort_order` is the record's cover, so "replace" is unambiguously
    a delete plus an add, and saying so in the payload keeps the diff readable.
    """

    id = serializers.IntegerField(required=False)
    image = serializers.CharField(required=False)

    def validate(self, attrs):
        has_id = attrs.get('id') is not None
        has_image = bool(attrs.get('image'))
        if has_id == has_image:
            raise serializers.ValidationError(
                'Each photo must carry either an "id" (keep this row) or an '
                '"image" (add this one), and not both.'
            )
        if has_image:
            sub = ImageProcessingSerializer(
                data={'base64_image': attrs['image']}, **photo_cfg()
            )
            if not sub.is_valid():
                raise serializers.ValidationError(
                    {'image': sub.errors['base64_image']}
                )
        return attrs


def photos_patch_field(required=True):
    """The `photos` field on an **edit**: the gallery as it should be afterwards.

    ⚠ **The list is the whole gallery, not the additions to it.** A row whose id
    is absent is *deleted*, a bare `image` is created, and `sort_order` is simply
    the index in this array - which is what makes re-ordering and changing the
    cover free rather than three more endpoints. `photos[0]` is the cover here
    exactly as it is on create (see `photos_field`).

    That contract is the reason the edit form can keep the create form's single
    Submit button. The alternative - add, delete and reorder as separate calls,
    which is what the CMS's gallery editor does - is right for an author working
    through a long list, and wrong for a contributor fixing one blurry photo on a
    phone: it turns one intention into three requests that can each half-fail.

    Omitting the field entirely leaves the gallery untouched, which is what a
    PATCH that only edits the prose sends.
    """
    return serializers.ListField(
        child=PhotoPatchItem(),
        required=required,
        allow_empty=not required,
        min_length=1 if required else 0,
        max_length=MAX_CONTRIBUTION_PHOTOS,
    )


class ContributionUpdateSerializer(serializers.ModelSerializer):
    """Base for the three public **edit** serializers.

    A sibling of both the CMS write serializer and the contribute *create*
    serializer, for the reason the module docstring gives: a field list that is
    inherited is a field list that can widen by accident, and this one is reached
    by any signed-in account.

    A subclass declares its own `Meta.fields`, sets
    `photo_write_serializer_class`, and overrides `photo_rows` when its gallery
    is not `instance.images` (a sighting's is `media`, which also holds clips).
    """

    #: The gallery writer for this record's photo table - the same one the create
    #: serializer uses, so both paths share the CMS's image pipeline.
    photo_write_serializer_class = None

    def photo_rows(self, instance):
        """The gallery rows this diff may delete. Photos only, never clips."""
        return instance.images.all()

    def _sync_photos(self, instance, photos):
        """Make the stored gallery match `photos`, in that order.

        Three passes, in this order and for a reason: the ids are resolved
        against **this record's own rows** first (so an id belonging to somebody
        else's gallery is a validation error rather than a silent no-op), the
        rows that survived are renumbered, and only then is anything deleted.
        """
        existing = {row.id: row for row in self.photo_rows(instance)}

        keep_ids = [item['id'] for item in photos if item.get('id') is not None]
        unknown = [pk for pk in keep_ids if pk not in existing]
        if unknown:
            raise serializers.ValidationError(
                {'photos': f'No such photo on this record: {unknown[0]}.'}
            )

        writer = self.photo_write_serializer_class
        for index, item in enumerate(photos):
            pk = item.get('id')
            if pk is not None:
                row = existing[pk]
                # `sort_order` is the cover, so this is the whole of what a
                # re-order writes - see `core.serializers.gallery_image_url`.
                if row.sort_order != index:
                    row.sort_order = index
                    row.save(update_fields=['sort_order'])
                continue

            sub = writer(data={'image': item['image'], 'sort_order': index})
            # Already validated by `PhotoPatchItem`; this pass is the writer's
            # own field coercion, not a re-check of the payload.
            sub.is_valid(raise_exception=True)
            sub.save(instance)

        for pk, row in existing.items():
            if pk not in set(keep_ids):
                row.delete()

    def _request_user(self):
        request = self.context.get('request')
        user = getattr(request, 'user', None)
        return user if (user is not None and user.is_authenticated) else None

    @transaction.atomic
    def update(self, instance, validated_data):
        photos = validated_data.pop('photos', None)

        for field, value in validated_data.items():
            setattr(instance, field, value)

        # ⚠ The edit itself un-publishes. A contributor may correct a record that
        # an administrator has already put on the site, and the correction has
        # not been reviewed - so it goes back into the queue rather than straight
        # onto the public page. `was_published` stays latched (see
        # `core.models.stamp_published`), which is what lets the frontend say
        # "back in review" instead of "awaiting first review".
        instance.enabled = False
        instance.save()

        if photos is not None:
            self._sync_photos(instance, photos)
        return instance

File: apps/animals/lib/contributions.ts
/**
 * The browser half of "My contributions" - what a signed-in reader has filed,
 * and the only place they may correct or withdraw it.
 *
 * The other direction of `lib/contribute.ts`. That module deliberately had two
 * functions and a note saying it should stay that way, because "a contribute
 * update endpoint would mean a pending record editable after review had
 * started". This is that decision revisited rather than ignored, and the thing
 * that makes it safe is on the API side: an edit **un-publishes** the record
 * (animals-api's `ContributionUpdateSerializer`), so a correction to something
 * already on the site goes back into the review queue instead of going live
 * unreviewed. See `CONTRIBUTION_STATUSES` below.
 *
 * Every call goes to the same-origin `/api/contributions/*` route handlers,
 * which attach the bearer token from the HTTP-only cookie and refresh it on a
 * 401 - so, exactly as in `lib/contribute.ts` and `lib/admin-api.ts`, there is
 * no token for this module to hold.
 */

import { ContributeError } from "@/lib/contribute";

/**
 * The three record types the contribute flow can create. The values are the URL
 * segments animals-api uses (`/api/contributions/<type>/<pk>/`) and this app's
 * own detail route (`/contributions/<type>/<id>`), so one string serves both.
 */
export const CONTRIBUTION_TYPES = ["sightings", "species", "locations"] as const;
export type ContributionType = (typeof CONTRIBUTION_TYPES)[number];

/** What one row calls itself on a payload - the singular of the type above. */
export type ContributionKind = "sighting" | "species" | "location";

/** The URL segment for a payload's `type`. The two differ only for a sighting. */
export const TYPE_FOR_KIND: Record<ContributionKind, ContributionType> = {
  sighting: "sightings",
  species: "species",
  location: "locations",
};

export function isContributionType(
  value: string | undefined,
): value is ContributionType {
  return CONTRIBUTION_TYPES.includes(value as ContributionType);
}

/**
 * The three states a contribution can be in.
 *
 * ⚠ `pending` and `in_review` are the **same row state** on the API - both are
 * `enabled: false` - and are told apart by `was_published`. The distinction is
 * the whole reason that column exists, and it is not cosmetic:
 *
 * - `pending` - filed, never yet published. Nothing has been lost by waiting.
 * - `published` - live on the public site.
 * - `in_review` - was live; its author edited it, so it has come **off** the
 *   public site until the edit is approved.
 *
 * A reader has to be told the third apart from the first, or editing a
 * published entry looks like it silently deleted their page.
 */
export const CONTRIBUTION_STATUSES = [
  "pending",
  "published",
  "in_review",
] as const;
export type ContributionStatus = (typeof CONTRIBUTION_STATUSES)[number];

export function isContributionStatus(
  value: string | undefined,
): value is ContributionStatus {
  return CONTRIBUTION_STATUSES.includes(value as ContributionStatus);
}

/**
 * One tile in the contributions grid.
 *
 * Deliberately **not** the record's full read payload: a grid of 24 tiles does
 * not need 24 galleries and 24 descriptions, so animals-api hand-builds this
 * shape (`core/my_contributions.py` → `_card`). The per-type fields below are
 * optional for that reason rather than because they are sometimes missing - a
 * sighting always has `date`, a species never does.
 */
export interface ContributionCard {
  type: ContributionKind;
  id: number;
  slug: string;
  name: string | null;
  en_name: string | null;
  status: ContributionStatus;
  enabled: boolean;
  was_published: boolean;
  created: string;
  modified: string;
  image: string | null;
  icon?: string | null;
  /** A sighting's day. `null` on the two catalog records. */
  date: string | null;

  // Sighting only.
  species_name?: string | null;
  species_en_name?: string | null;
  location_name?: string | null;
  location_en_name?: string | null;
  has_video?: boolean;

  // Species only.
  scientific_name?: string | null;
  category_name?: string | null;
  category_en_name?: string | null;

  // Location only.
  place_type?: string | null;
  county_name?: string | null;
}

export interface ContributionPage {
  count: number;
  limit: number;
  offset: number;
  results: ContributionCard[];
}

/**
 * A record's full read payload, as the edit form prefills from.
 *
 * It **is** the record's normal read shape - the same one every other read of
 * that resource gives - plus the three contribution fields. Typed loosely
 * because the three records have genuinely different shapes and each form reads
 * only its own; the alternative is a discriminated union repeating three
 * payloads that `lib/catalog.ts` and `lib/journal.ts` already describe.
 */
export type ContributionDetail = Record<string, unknown> & {
  id: number;
  slug: string;
  type: ContributionKind;
  contribution_status: ContributionStatus;
  was_published: boolean;
};

/**
 * One entry in a `photos` edit: keep a stored row, or add a new photograph.
 *
 * ⚠ **The list is the gallery afterwards, not the additions to it** - a stored
 * row left out is deleted, and `sort_order` is the index in the array, so
 * `photos[0]` is the cover. See animals-api's `photos_patch_field`; it is what
 * lets the edit form keep the filing form's single Submit button instead of
 * turning one intention into three requests that can each half-fail.
 */
/**
 * What turns one of the three contribute forms from a filing form into an edit
 * form.
 *
 * The forms are shared rather than duplicated because a contributor correcting a
 * sighting should meet the form they filled in, not a different one - which was
 * the explicit brief for this page. So each form takes this prop plus its own
 * `initial…` values, and everything else about it is unchanged: the same stages,
 * the same fields, the same validation.
 *
 * Absent, the form files a new record. Present, it PATCHes this one.
 */
export interface ContributionEdit {
  id: number;
  /** Where the record stands *now* - what decides whether to warn before saving. */
  status: ContributionStatus;
  /**
   * Called after a successful save, with the status the record now has.
   *
   * ⚠ It is passed the status **from the response**, not the one above: saving a
   * `published` record makes it `in_review`, and the page has to be able to say
   * so. Never assume the status survived the save.
   */
  onSaved: (status: ContributionStatus) => void;
}

export type PhotoPatch = { id: number } | { image: string };

/**
 * The `photos` diff for a picker's current tiles, in their current order.
 *
 * A tile that came off the record keeps its row (`{id}`); one the contributor
 * just picked is sent as a data URL. So an edit that only re-orders uploads
 * nothing at all - it sends three ids in a new order - which is the difference
 * between changing a cover instantly and re-uploading a whole gallery over
 * cellular data.
 */
export function photoPatch(
  photos: { id?: number; dataUrl: string }[],
): PhotoPatch[] {
  return photos.map((photo) =>
    photo.id !== undefined ? { id: photo.id } : { image: photo.dataUrl },
  );
}

/**
 * A record's stored gallery as picker tiles.
 *
 * `dataUrl` holds the **stored** URL here rather than a data URL - the picker
 * only ever renders it, and `PickedPhoto.id` is what tells the two apart when
 * the diff is built. `name` falls back to the record's own so the alt text is
 * never empty; a gallery row's caption is optional and usually absent.
 */
export function galleryAsPhotos(
  rows: { id: number; image: string | null; name?: string | null }[] | undefined,
  fallbackName: string,
): { key: string; dataUrl: string; name: string; id: number }[] {
  return (rows ?? [])
    .filter((row): row is typeof row & { image: string } => Boolean(row.image))
    .map((row) => ({
      key: `stored-${row.id}`,
      dataUrl: row.image,
      name: row.name?.trim() || fallbackName,
      id: row.id,
    }));
}

async function request<T>(path: string, init: RequestInit = {}): Promise<T> {
  const res = await fetch(`/api/contributions/${path}`, {
    headers: { "Content-Type": "application/json" },
    ...init,
  });

  if (!res.ok) {
    const data: Record<string, unknown> = await res.json().catch(() => ({}));
    // The same error type the filing flow raises, so `firstErrorMessage` works
    // on both and a form does not need two catch branches.
    throw new ContributeError(res.status, data);
  }
  // 204 from a withdrawal - there is no body to parse.
  if (res.status === 204) return undefined as T;
  return res.json() as Promise<T>;
}

/** Everything this account has filed, newest first across all three types. */
export function listContributions(params: {
  type?: ContributionType;
  status?: ContributionStatus;
  limit?: number;
  offset?: number;
} = {}): Promise<ContributionPage> {
  const query = new URLSearchParams();
  if (params.type) query.set("type", params.type);
  if (params.status) query.set("status", params.status);
  if (params.limit !== undefined) query.set("limit", String(params.limit));
  if (params.offset !== undefined) query.set("offset", String(params.offset));

  const suffix = query.toString();
  return request<ContributionPage>(suffix ? `?${suffix}` : "");
}

/** One record, in the shape its edit form prefills from. */
export function getContribution(
  type: ContributionType,
  id: number,
): Promise<ContributionDetail> {
  return request<ContributionDetail>(`${type}/${id}`);
}

/**
 * Correct a record.
 *
 * ⚠ **This un-publishes a published record.** The response's
 * `contribution_status` is what actually happened - read it rather than assuming
 * the record stayed as it was.
 */
export function updateContribution(
  type: ContributionType,
  id: number,
  body: Record<string, unknown>,
): Promise<ContributionDetail> {
  return request<ContributionDetail>(`${type}/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body),
  });
}

/**
 * Withdraw a record.
 *
 * May be refused with a **409**: a published species that other people's
 * sightings now reference is protected on the API side, which is what stops one
 * withdrawal taking somebody else's journal entries with it.
 */
export function deleteContribution(
  type: ContributionType,
  id: number,
): Promise<void> {
  return request<void>(`${type}/${id}`, { method: "DELETE" });
}
